smst_test.go

```diff
@@ -359,74 +359,108 @@ func TestSMST_OrphanRemoval(t *testing.T) {
 		err = smst.Update([]byte("testKey"), []byte("testValue"), 5)
 		require.NoError(t, err)
 		require.Equal(t, 1, nodeCount(t)) // only root node
+		require.Equal(t, uint64(1), impl.Count())
 	}
 
 	t.Run("delete 1", func(t *testing.T) {
 		setup()
 		err = smst.Delete([]byte("testKey"))
 		require.NoError(t, err)
 		require.Equal(t, 0, nodeCount(t))
+		require.Equal(t, uint64(0), impl.Count())
 	})
 
 	t.Run("overwrite 1", func(t *testing.T) {
 		setup()
 		err = smst.Update([]byte("testKey"), []byte("testValue2"), 10)
 		require.NoError(t, err)
 		require.Equal(t, 1, nodeCount(t))
+		require.Equal(t, uint64(1), impl.Count())
 	})
 
-	type testCase struct {
-		keys  []string
-		count int
-	}
-	// sha256(testKey)  = 0001...
-	// sha256(testKey2) = 1000... common prefix len 0; 3 nodes (root + 2 leaf)
-	// sha256(foo)      = 0010... common prefix len 2; 5 nodes (3 inner + 2 leaf)
-	cases := []testCase{
-		{[]string{"testKey2"}, 3},
-		{[]string{"foo"}, 4},
-		{[]string{"testKey2", "foo"}, 6},
-		{[]string{"a", "b", "c", "d", "e"}, 14},
-	}
-
 	t.Run("overwrite and delete", func(t *testing.T) {
 		setup()
 		err = smst.Update([]byte("testKey"), []byte("testValue2"), 2)
 		require.NoError(t, err)
 		require.Equal(t, 1, nodeCount(t))
+		require.Equal(t, uint64(1), impl.Count())
 
 		err = smst.Delete([]byte("testKey"))
 		require.NoError(t, err)
 		require.Equal(t, 0, nodeCount(t))
+		require.Equal(t, uint64(0), impl.Count())
+	})
 
-		for tci, tc := range cases {
+	type testCase struct {
+		desc              string
+		keys              []string
+		expectedNodeCount int
+		expectedLeafCount int
+	}
+	// sha256(testKey)  = 0001...
+	// sha256(testKey2) = 1000... common prefix len 0; 3 nodes (root + 2 leaf)
+	// sha256(foo)      = 0010... common prefix len 2; 5 nodes (3 inner + 2 leaf)
+	cases := []testCase{
+		{
+			desc:              "insert a single key (testKey2) which DOES HAVE a similar prefix to the already present key (testKey)",
+			keys:              []string{"testKey2"},
+			expectedNodeCount: 3,
+			expectedLeafCount: 2,
+		},
+		{
+			desc:              "insert a single key (foo) which DOES NOT HAVE a similar prefix to the already present key (testKey)",
```

Comment:
I think the top one should say `DOES NOT HAVE` and the second `DOES HAVE` as the comment suggests
```
sha256(testKey)=0001...
sha256(testKey2)=1000... // 0 bits common prefix
sha256(foo)=0010... // 2 bits common prefix
```
So surely the descriptions are backwards? Because foo will follow the same path as testKey making an extension node with an inner node child but testKey2 will automatically branch to the right making just an inner node